File: app/src/main/java/com/example/trialproject3/Fragment/CartFragment.java
package com.example.trialproject3.Fragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.trialproject3.Activity.AddressActivity;
import com.example.trialproject3.Activity.CartActivity;
import com.example.trialproject3.Activity.MainActivity;
import com.example.trialproject3.Activity.ToShipActivity;
import com.example.trialproject3.Adapter.CartListAdapter;
import com.example.trialproject3.Helper.ChangeNumberItemsListener;
import com.example.trialproject3.Helper.ManagementCart;
import com.example.trialproject3.Map.MapboxMapActivity;
import com.example.trialproject3.R;
import com.example.trialproject3.databinding.FragmentCartBinding;

import java.io.Serializable;

public class CartFragment extends Fragment implements MainActivity.OnBackPressedListener {
    private static final String TAG = "CartFragment";
    private FragmentCartBinding binding;
    private Context context;
    private RecyclerView.Adapter adapter;
    private ManagementCart managementCart;
    private double tax;
    private AlertDialog.Builder builder;
    private Intent intent;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentCartBinding.inflate(inflater, container, false);

        context = requireContext();
        builder = new AlertDialog.Builder(context);
        binding.emptyTxt.setVisibility(View.GONE);

        intent = requireActivity().getIntent();
        if (intent != null && intent.hasExtra("selectedAddress")) {
            String selectedAddress = intent.getStringExtra("selectedAddress");
            binding.addressTextView.setText(selectedAddress);
        }

        binding.orderNowBtn.setOnClickListener(v -> builder.setTitle("Are you sure?")
                .setMessage("Do you want to purchase these items?")
                .setCancelable(true)
                .setPositiveButton("Yes", (dialog, which) -> {
                    Intent intent1 = requireActivity().getIntent();
                    String selectedAddress = "";
                    if (intent1 != null && intent1.hasExtra("selectedAddress")) {
                        selectedAddress = intent1.getStringExtra("selectedAddress");
                    }

                    Intent toShipIntent = new Intent(context, ToShipActivity.class);
                    toShipIntent.putExtra("selectedAddress", selectedAddress);
                    toShipIntent.putExtra("cartItems", (Serializable) managementCart.getListCart());
                    startActivity(toShipIntent);

                    requireActivity().finish();
                })
                .setNegativeButton("No", (dialog, which) -> dialog.cancel())
                .show());


//        toProfileBtn.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                List<PopularDomain> cartItems = managementCart.getListCart();
//                Intent profileIntent = new Intent(CartActivity.this, ProfileActivity.class);
//                profileIntent.putExtra("cartItems", (Serializable) cartItems);
//                startActivity(profileIntent);
//            }
//        });

        binding.backBtn.setOnClickListener(v -> builder.setTitle("Are you sure?")
                .setMessage("Do you want to leave your cart?")
                .setCancelable(true)
                .setPositiveButton("Yes", (dialog, which) -> {
                    dialog.cancel();
                    backToHomeFragment();
                })
                .setNegativeButton("No", (dialog, which) -> dialog.cancel())
                .show());

        binding.addAddressBtn.setOnClickListener(v ->
                startActivity(new Intent(context, AddressActivity.class)));

        managementCart = new ManagementCart(context);

        initView();
        initList();
        calculateCart();

        return binding.getRoot();

    }
    @Override
    public boolean onBackPressed() {
        backToHomeFragment();
        return true;
    }
    private void initList() {
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false);
        binding.itemsRecyclerView.setLayoutManager(linearLayoutManager);
        adapter = new CartListAdapter(managementCart.getListCart(), context, new ChangeNumberItemsListener() {
            @Override
            public void change() {
                calculateCart();
            }
        });

        binding.itemsRecyclerView.setAdapter(adapter);
        if (managementCart.getListCart().isEmpty()) {
            binding.emptyTxt.setVisibility(View.VISIBLE);
            binding.scrollView.setVisibility(View.GONE);
        } else {
            binding.emptyTxt.setVisibility(View.GONE);
            binding.scrollView.setVisibility(View.VISIBLE);
        }
    }

    private void backToHomeFragment() {
        FragmentManager fragmentManager = requireActivity().getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(R.id.fragmentContainer, new HomeFragment());
        fragmentTransaction.addToBackStack(null);
        fragmentTransaction.commit();
    }

    @SuppressLint("SetTextI18n")
    private void calculateCart() {
        double percentTax = 0.02;
        double delivery = 10;
        tax = Math.round((managementCart.getTotalFee() * percentTax * 100.0)) / 100.0;

        double total = Math.round((managementCart.getTotalFee() + tax + delivery) * 100.0) / 100.0;
        double itemTotal = Math.round(managementCart.getTotalFee() * 100) / 100;

        binding.totalFeeTxt.setText("₱" + itemTotal);
        binding.taxTxt.setText("₱" + tax);
        binding.deliveryTxt.setText("₱" + delivery);
        binding.totalTxt.setText("₱" + total);
    }

    private void initView() {
        if (managementCart.getListCart().isEmpty()) {
            binding.backBtn.setVisibility(View.VISIBLE);
        } else {
            binding.backBtn.setVisibility(View.VISIBLE);
        }
    }


}

File: app/src/main/java/com/example/trialproject3/Fragment/CartFragment.kt
package com.example.trialproject3.Fragment

import android.annotation.SuppressLint
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RadioButton
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.trialproject3.Activity.AddressActivity
import com.example.trialproject3.Activity.MainActivity.OnBackPressedListener
import com.example.trialproject3.Adapter.CartAdapter
import com.example.trialproject3.Firebase.FirebaseHelper
import com.example.trialproject3.Helper.AlertDialogHelper
import com.example.trialproject3.Models.CartModel
import com.example.trialproject3.Utility.FragmentManagerHelper
import com.example.trialproject3.Utility.LoadingSpinnerOverlay
import com.example.trialproject3.Utility.ToastHelper
import com.example.trialproject3.databinding.FragmentCartBinding

class CartFragment : Fragment(),
    OnBackPressedListener,
    CartAdapter.OnIncreaseCartClickListener,
    CartAdapter.OnDecreaseCartClickListener,
    CartAdapter.OnRemoveCartItemClickListener {

    private lateinit var binding: FragmentCartBinding
    private lateinit var context: Context
    private lateinit var builder: AlertDialog.Builder
    private lateinit var toastHelper: ToastHelper
    private lateinit var cartAdapter: CartAdapter
    private lateinit var loadingSpinnerOverlay: LoadingSpinnerOverlay
    private lateinit var alertDialogHelper: AlertDialogHelper
    private lateinit var fragmentManagerHelper: FragmentManagerHelper
    private var selectedPaymentMethod: String? = null

    companion object {
        private const val TAG = "CartFragment"
    }

    @SuppressLint("SetTextI18n")
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentCartBinding.inflate(inflater, container, false)

        context = requireContext()
        builder = AlertDialog.Builder(context)
        toastHelper = ToastHelper(context)
        loadingSpinnerOverlay = LoadingSpinnerOverlay(context)
        alertDialogHelper = AlertDialogHelper(context)
        fragmentManagerHelper = FragmentManagerHelper(activity as FragmentActivity)

        binding.backBtn.setOnClickListener { fragmentManagerHelper.showFragment(HomeFragment()) }

        binding.paymentMethodRadioGroup.setOnCheckedChangeListener { group, checkedId ->
            val radioButton = group.findViewById<RadioButton>(checkedId)
            selectedPaymentMethod = radioButton.text.toString()
            binding.paymentMethodTextView.text = "Payment Method: $selectedPaymentMethod"
        }

        binding.shopBtn.setOnClickListener { fragmentManagerHelper.showFragment(HomeFragment()) }

        binding.addAddressBtn.setOnClickListener {
            startActivity(
                Intent(
                    context,
                    AddressActivity::class.java
                )
            )
        }

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (isAdded) loadCart()
    }

    override fun onBackPressed(): Boolean {
        fragmentManagerHelper.showFragment(HomeFragment())
        return true
    }

    @SuppressLint("SetTextI18n")
    private fun loadCart() {
        loadingSpinnerOverlay.showLoading()
        val cartReference = FirebaseHelper.getFireStoreInstance()
            .collection(FirebaseHelper.KEY_COLLECTION_CARTS)

        cartReference.addSnapshotListener { value, error ->
            if (error != null) {
                Log.e(TAG, "loadCart: " + error.message)
                loadingSpinnerOverlay.hideLoading()
                return@addSnapshotListener
            }

            if (value != null && !value.isEmpty) {
                val cartsList = mutableListOf<CartModel>()
                for (carts in value.documents) {
                    val cartModel = carts.toObject(CartModel::class.java)
                    if (cartModel != null && cartModel.cartUserID == FirebaseHelper.currentUserID()) {
                        binding.emptyCartLayout.visibility = View.GONE
                        if (cartModel.cartUserID == FirebaseHelper.currentUserID()) {
                            cartsList.add(cartModel)

                            binding.orderNowBtn.setOnClickListener {
                                if (selectedPaymentMethod != null)
                                    builder.setTitle("Are you sure?")
                                        .setMessage("Do you want to purchase these items?")
                                        .setCancelable(true)
                                        .setPositiveButton("Yes") { dialog: DialogInterface?, which: Int ->
                                            decreaseStock(cartAdapter.cartModelList)
                                        }
                                        .setNegativeButton("No") { dialog: DialogInterface, which: Int -> dialog.cancel() }
                                        .show()
                                else toastHelper.showToast(
                                    "Please select a payment method",
                                    1
                                )

                            }
                        }
                    }
                }
                if (cartsList.isNotEmpty()) {
                    cartAdapter =
                        CartAdapter(
                            context,
                            cartsList,
                            this@CartFragment,
                            this@CartFragment,
                            this@CartFragment
                        )
                    binding.cartRecyclerView.layoutManager = LinearLayoutManager(context)
                    binding.cartRecyclerView.adapter = cartAdapter

                    val subtotal = cartAdapter.calculateTotalPrice()
                    binding.subTotalTextView.text = "Subtotal: ₱$subtotal"

                    val totalPrice = subtotal
                    binding.totalTextView.text = "Total Price: ₱$totalPrice"
                } else {
                    binding.cartLayout.visibility = View.GONE
                }
            } else {
                binding.cartLayout.visibility = View.GONE
            }
            loadingSpinnerOverlay.hideLoading()
        }
    }

    override fun onIncreaseCartClick(cartModel: CartModel) {
        val productReference = FirebaseHelper.getFireStoreInstance()
            .collection(FirebaseHelper.KEY_COLLECTION_PRODUCTS)
            .document(cartModel.productID)

        productReference.get()
            .addOnSuccessListener { document ->
                if (document != null) {
                    val currentStock = document.getLong("stock") ?: 0
                    if (cartModel.quantity < currentStock) {
                        cartModel.quantity++
                        updateQuantity(cartModel)
                    } else {
                        toastHelper.showToast("Cannot add more items. Product is out of stock.", 1)
                    }
                } else {
                    Log.e(TAG, "No such document")
                }
            }
            .addOnFailureListener { exception ->
                Log.e(TAG, "Failed to get document: ${exception.message}")
            }
    }

    override fun onDecreaseCartClick(cartModel: CartModel) {
        if (cartModel.quantity > 1) {
            cartModel.quantity--
            updateQuantity(cartModel)
        }
    }

    override fun onRemoveCartItemClick(cartModel: CartModel) {
        val cartReference = FirebaseHelper.getFireStoreInstance()
            .collection(FirebaseHelper.KEY_COLLECTION_CARTS)
            .document(cartModel.cartID)

        alertDialogHelper.showAlertDialog(
            "Cart",
            "Are you sure you want to remove this item from your cart?",
            "Remove",
            { dialog, id ->
                cartReference.delete()
                    .addOnSuccessListener {
                        Log.d(TAG, "DocumentSnapshot successfully deleted!")
                    }
                    .addOnFailureListener { e ->
                        Log.w(TAG, "Error deleting document", e)
                    }
            },
            "Close",
            { dialog, id ->
                dialog.dismiss()
            }
        )
    }

    @SuppressLint("NotifyDataSetChanged")
    private fun updateQuantity(cartModel: CartModel) {
        val cartReference = FirebaseHelper.getFireStoreInstance()
            .collection(FirebaseHelper.KEY_COLLECTION_CARTS)
            .document(cartModel.cartID)

        cartReference.update("quantity", cartModel.quantity)
            .addOnSuccessListener {
                if (cartAdapter.cartModelList.isEmpty()) {
                    binding.emptyCartLayout.visibility = View.VISIBLE
                } else {
                    binding.emptyCartLayout.visibility = View.GONE
                }
                cartAdapter.notifyDataSetChanged()
            }
            .addOnFailureListener { e ->
                Log.e(TAG, "Failed to update quantity: ${e.message}")
            }
    }

    private fun decreaseStock(cartModels: List<CartModel>) {
        loadingSpinnerOverlay.showLoading()
        for (cartModel in cartModels) {
            val productReference = FirebaseHelper.getFireStoreInstance()
                .collection(FirebaseHelper.KEY_COLLECTION_PRODUCTS)
                .document(cartModel.productID)

            productReference.get().addOnSuccessListener { document ->
                if (document != null) {
                    val currentStock = document.getLong("stock") ?: 0
                    if (currentStock >= cartModel.quantity) {
                        val newStock = currentStock - cartModel.quantity
                        productReference.update("stock", newStock)
                            .addOnCompleteListener { task ->
                                loadingSpinnerOverlay.hideLoading()
                                if (task.isSuccessful) {
                                    fragmentManagerHelper.showFragment(OrderOnTheWayFragment())
                                    Log.d(TAG, "Stock decreased successfully for product: ${cartModel.productID}")
                                } else {
                                    Log.e(TAG, "Failed to decrease stock: ${task.exception?.message}")
                                }
                            }
                    } else {
                        Log.e(TAG, "Not enough stock available for product: ${cartModel.productID}")
                    }
                } else {
                    Log.e(TAG, "No such document for product: ${cartModel.productID}")
                }
            }.addOnFailureListener { exception ->
                Log.e(TAG, "Failed to get document: ${exception.message}")
            }
        }
    }}